### DISCUSSION TITLE
CP Editor sucks

### DISCUSSION THREAD

**Discussion by ouuan** [cpeditor/cpeditor#760] [Show and tell]
**CP Editor sucks.**

I admit this in spite of all the great features CP Editor has.

If you also think CP Editor sucks, feel free to express any of your thoughts here:

-   Small troubles when using CP Editor.
-   Missing features that make CP Editor useless.
-   Non-intuitive designs that are really confusing.
-   ...

<details>
<summary>How this post is special</summary>
<br />

If you open an issue:
1. We require you to fill in the issue template to provide enough information and details.
2. We try our best to give instant replies and solve your problem.
3. We'll be unhappy if the problem you met turns out to be your own fault.
4. You shouldn't request the same feature / report the same bug again if there's already another issue for it.

You can also ask questions in [Discussions](https://github.com/cpeditor/cpeditor/discussions/new), where we'll be happy to answer all kinds of questions.

But here in this post:
1. There's NO requirement at all. You can express your complaints as you like.
2. We'll read the comments, and try to improve CP Editor according to your complaints. But there's no guarantee that we'll respond to them.
</details>

**Comment by ShizuhaAki** [comment]
~~Ya know, I was kinda angry when I saw someone use such an insulting issue title as I checked my mail~~

I think a code block folding feature would be nice.

Yes, we are not working on huge projects here, but, imagine you have copied your template for a huge DS somewhere and it really annoys when you have to scroll all the way through the DS just to add a header/change a constant/...

**Reply by coder3101** [reply]
Just FYI, Code folding will be present in New Code Editor based CP Editor, in fact it is already available and works pretty good unless you use code linting.

**Reply by coder3101** [reply]
I do agree to your strike-through text. Do you have any alternate text for it?

**Reply by ouuan** [reply]
> I do agree to your strike-through text. Do you have any alternate text for it?

I think it's humorous and attracts people who really think so :thinking:

**Reply by coder3101** [reply]
Nay, it's just feels bad to read such title when we put so much efforts to it. But I do agree this is the best way to attract and engage with critics.

**Reply by tusqasi** [reply]
>  in fact it is already available and works pretty good unless you use code linting.

Where is it available? Is it not in fake-vim branch

**Reply by coder3101** [reply]
> >  in fact it is already available and works pretty good unless you use code linting.
> 
> Where is it available? Is it not in fake-vim branch

#516 but use this https://github.com/cpeditor/cpeditor/pull/516/commits/7317fd506974cba59243a42de40880db4fdaaade

**Comment by Arius1408** [comment]
Well, CP Editor does work great on linux, but when it come to macOS, it is kinda trouble that I can not use <bits/stdc++.h>. It there a way to do it ??.

**Reply by coder3101** [reply]
Sounds like a Q&A? Would you mind asking it in Q&A section of discussion?

**Reply by Ayx03** [reply]
You can save the headfile from [here](https://www.luogu.com.cn/paste/4tf6txv6) and manually add it to your library, then ```#include<bits/stdc++.h>``` will work fine

**Comment by Arius1408** [comment]
Ok, thank you for reply, and for for your great work

**Comment by tusqasi** [comment]
CP editor is great for what is. Sure it has a lot do.
autocomplete will be nice, with code folding as @Ravenclaw-OIer said.

**Comment by ameyajangam22** [comment]
Well I just had two more suggestions apart from the autocomplete feature .Whenever we try to match some characters like  
                                                     `if (s[i] ==' ( ')`   

The character inside the single quotes should also be highlighted . This would give better readability while checking code before submitting to the judge.  

A second feature which I think will be great if it gets added (I am not sure if it's possible or not) i.e adding AStyle Formatter as a formatting option . Clang is good but AStyle just feels better .

**Reply by tusqasi** [reply]
I think this is being worked on right now.

**Reply by ameyajangam22** [reply]
Oh that's great

**Comment by freeze-oi** [comment]
This debugging function is hard to use ; (

**Reply by freeze-oi** [reply]
I hope you can improve it

**Comment by tusqasi** [comment]
In vim mode (in fake-vim-support branch) there should be a way to add comments through a vim command. 

Toggle comment edit action is already in cp-editor. This can be added by exposing existing editing functions to vim commands.

I don't know how that would work, lemme know if that is even possible

**Reply by coder3101** [reply]
Yes that's possible and a nice feature to add, Usually in my neovim, I use plugin called [`Vim Commentary`](https://github.com/tpope/vim-commentary) that exposes a command `:Commentary` which toggles comments on a line on which cursor was present, or whole block of selection if some text is selected.

**Reply by tusqasi** [reply]
Yeah, i use a plugin for that in vim too.

**Reply by tusqasi** [reply]
But adding a plugin system might not be needed for this problem. 
This is a nice feature in it self though.

Commenting is already in the editor. Can't that function be exposed to the vim mode.

In those lines can all the functions of cpeditor be exposed to vim mode?

**Reply by coder3101** [reply]
> But adding a plugin system might not be needed for this problem. 
> This is a nice feature in it self though.

Yes, We need to embed an actual Neovim to have full vim support with extensions. Since fake-vim is a vim emulation, it is not feasible to add extension as they depend on n/vim api.
> 
> Commenting is already in the editor. Can't that function be exposed to the vim mode.
> 

Yes, it will be exposed as other functions like run, detach run are exposed. Plan is to eventually expose all the functions of cpeditor to vim.

**Comment by tusqasi** [comment]
Can treesitter be used for syntaxhighlighting? It works nice with neovim.

**Reply by coder3101** [reply]
> Can treesitter be used for syntaxhighlighting? It works nice with neovim.

Yes if you are willing to contribute please go ahead. The new editor is complete overhaul and will take much longer considering that we are not getting enough time for it lately.

**Reply by tusqasi** [reply]
I'll take a look at what i can do

**Comment by psycoAppDemon** [comment]
How to restore previous version of cp editor
Because I updated it today and it is not running the code correctly sometimes giving tle sometimes other output

**Reply by coder3101** [reply]
Whats your OS? What was your old version. Which version you updated to?

Can you share a concise and precise example of whats not working in updated version? 

PS: You should open this as Q&A not Complaint section. Your question is "How to rollback to old version of cpeditor?"

**Comment by psycoAppDemon** [comment]
This is one of the problems
Besides this it is giving different outputs when run the same code many
times or sometimes tle

**Reply by coder3101** [reply]
Can you try a minimal code example on which it is giving different output for same code and input. Sounds like an UB to me 🤔

**Comment by psycoAppDemon** [comment]
Okk go through my submission you will see wrong ans on tc 1 but when I run
the same code before submitting it gave tle and then the same code gave
correct ans and then sometimes gave runtime error finall the msg as per the
last ss I sent to you
https://codeforces.com/submissions/codingBeast_oo7#

**Reply by coder3101** [reply]
How your code behaves in codeforces is not cpeditor's business. Can you produce a minimal example that produces unexpected output locally? (Provided same code works fine with other tool)

**Comment by psycoAppDemon** [comment]
Os windows 10
I updated to cp editor to version 6.9.4

**Comment by psycoAppDemon** [comment]
No need to worry I will go for some other editors then!

**Comment by kevinyang07** [comment]
I am having trouble with the language server having problems with <bits/stdc++.h> and I would like to suggest adding vim

**Reply by michelececcacci** [reply]
I would love vim support, too

**Reply by Gr-Nobel** [reply]
yes.vim support.

**Reply by pavan-kalyan** [reply]
Yes please, lack of vim support is the only thing stopping me from using CP editor.

**Comment by coderbhawesh** [comment]
I am unable to use it for macOS. It says cannot verify the developer.
It says:
   “cpeditor” cannot be opened because the developer cannot be verified.
Please help!!

**Reply by ouuan** [reply]
https://cpeditor.org/docs/installation/download-from-releases/#macos

> add an exception to CP Editor from System Preferences -> Security & Privacy -> General, and allow CP Editor to run anyways

**Comment by hehaodele** [comment]
It seems CPeditor does not support cutting one line where the cursor is by a hotkey. I really want to use "Ctrl + X" to cut the line I am editing as I can do in visual studio, sublime text, pycharm...
Is there way to do it in CPeditor? do I just miss something?

**Reply by ouuan** [reply]
There is no cutting one line, but there are <kbd>Ctrl+Shift+K</kbd> to delete one line and <kbd>Ctrl+Shift+↑/↓</kbd> to move one line.

**Comment by ammardab3an** [comment]
How can I disable the (ctrl+scroll) feature, it's painfully slow, and I keep doing it by mistake, and it would take 5s to zoom in and out. For some reason to changing the font size takes too much time.

**Comment by Akagishigeruzzx** [comment]
Can you add the function of dragging code

**Comment by Akagishigeruzzx** [comment]
能加上拖动代码的功能吗？

**Comment by songbobobobo** [comment]
We need AStyle support!

**Comment by EricWan377873** [comment]
想要拖动代码，想要自定义渲染格式。

**Comment by Daniel-Dami** [comment]
I can press Control + / to comment a line in the editor even when my cursor is not focused on the editor.

https://github.com/user-attachments/assets/63b052af-f52f-47b7-a484-1795468b47a1

Like in the video, when I'm editing the testcases, the code still gets commented.
